### PR TITLE
Limit the number of PDUs/EDUs per fedreation transaction

### PR DESCRIPTION
**Pull request by erikjohnston**
As per spec: https://matrix.org/speculator/spec/HEAD/server_server/unstable.html#transactions

### CODE REVIEW

File: changelog.d/3805.misc [hunk 1]
```diff
@@ -0,0 +1 @@
+Limit the number of PDUs/EDUs per fedreation transaction
```
Comment (hawkowl) [inline changelog.d/3805.misc]: s/fedreation/federation/

File: synapse/federation/transaction_queue.py [hunk 2]
```diff
@@ -463,7 +463,19 @@ def _transaction_transmission_loop(self, destination):
                 # pending_transactions flag.
 
                 pending_pdus = self.pending_pdus_by_dest.pop(destination, [])
+
+                # We can only include at most 50 PDUs per transactions
+                pending_pdus, leftover_pdus = pending_pdus[:50], pending_pdus[50:]
+                if leftover_pdus:
+                    self.pending_pdus_by_dest[destination] = leftover_pdus
```
Comment (hawkowl) [inline synapse/federation/transaction_queue.py]: This makes it non-atomic -- it should be `self.pending_pdus_by_dest[destination].extend(leftover_pdus)`, maybe?

Comment (erikjohnston) [reply synapse/federation/transaction_queue.py]: I think by "atomic" it means "doesn't yield", tbh

Comment (hawkowl) [reply synapse/federation/transaction_queue.py]: if potential threading isn't a concern, then, looks fine to me.

Comment (richvdh) [reply synapse/federation/transaction_queue.py]: > I think by "atomic" it means "doesn't yield", tbh

this, ftr

**Pull request merged**